sdk/language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -78,45 +78,52 @@ optionsParserInfo = info (optionsParser <**> helper)
+    darToPackages :: FilePath -> IO [(PackageId, (Package, PackageReference))]
     darToPackages dar = do
       dar <- B.readFile dar
       let archive = Zip.toArchive $ BSL.fromStrict dar
       dalfs <- either fail pure $ DAR.readDalfs archive
       DAR.DalfManifest{packageName} <- either fail pure $ DAR.readDalfManifest archive
-      packageName <- pure (PackageName . T.pack <$> packageName)
-      forM ((DAR.mainDalf dalfs, packageName) : map (, Nothing) (DAR.dalfs dalfs)) $
-        \(dalf, mbPkgName) -> do
+      let mbPkgNmVer = readPackageNameAndVersion <$> packageName
+      forM ((DAR.mainDalf dalfs, mbPkgNmVer) : map (, Nothing) (DAR.dalfs dalfs)) $
+        \(dalf, mbPkgNmVer) -> do
           (pkgId, pkg) <- either (fail . show)  pure $ Archive.decodeArchive Archive.DecodeAsMain (BSL.toStrict dalf)
-          pure (pkgId, (pkg, mbPkgName))
+          let pkgRef = maybe (PkgId pkgId) PkgNameVer mbPkgNmVer
+          pure (pkgId, (pkg, pkgRef))
 
 -- Work out the set of packages we have to generate script for and by
 -- what names.
-mergePackageMap :: [(PackageId, (Package, Maybe PackageName))] ->
-                   Either T.Text (Map.Map PackageId (Maybe PackageName, Package))
+mergePackageMap :: [(PackageId, (Package, PackageReference))] ->
+                   Either T.Text (Map.Map PackageId (PackageReference, Package))
 mergePackageMap ps = foldM merge Map.empty ps
   where
-    merge :: Map.Map PackageId (Maybe PackageName, Package) ->
-                  (PackageId, (Package, Maybe PackageName)) ->
-                  Either T.Text (Map.Map PackageId (Maybe PackageName, Package))
-    merge pkgs (pkgId, (pkg, mbPkgName)) = do
-        let pkgNames = mapMaybe fst (Map.elems pkgs)
+    merge :: Map.Map PackageId (PackageReference, Package) ->
+                  (PackageId, (Package, PackageReference)) ->
+                  Either T.Text (Map.Map PackageId (PackageReference, Package))
+    merge pkgs (pkgId, (pkg, pkgRef)) = do
+        let pkgRefs = map fst (Map.elems pkgs)
         -- Check if there is a package with the same name but a
         -- different package id.
-        whenJust mbPkgName $ \name -> when (pkgId `Map.notMember` pkgs && name `elem` pkgNames) $
-            Left $ "Duplicate name '" <> unPackageName name <> "' for different packages detected"
+        when (pkgId `Map.notMember` pkgs && pkgRef `elem` pkgRefs) $
+            Left $ "Duplicate name '" <> forFilePath pkgRef <> "' for different packages detected"
         let update mbOld = case mbOld of
-                Nothing -> pure (Just (mbPkgName, pkg))
-                Just (mbOldPkgName, _) -> do
+                Nothing -> pure $ Just (pkgRef, pkg)
+                Just (oldPkgRef, _) -> do
                     -- Check if we have colliding names for the same
                     -- package.
-                    whenJust (liftA2 (,) mbOldPkgName mbPkgName) $ \(name1, name2) ->
```

Comment:
We can remove this layer of conditionals as both old and new will now be non-`Nothing`. A missing package name+version would have meant that this resolved to a `PkgId` on line 102, i.e.

https://github.com/digital-asset/daml/pull/19321/files#diff-db5da09d621f0cd3bf1c9e237a81922907961d0bb1df0d85405058c705be1653R102